### PR TITLE
Rework clock muxes into a chain structure

### PR DESCRIPTION
**Pull request by lumag**
Enforcing a primary+leaf configuration does not walk for all the cases. Some clocks (msm8996 CPU) have more muxes to be powered on. Other (MCCC) do not require GCC at all. Make clock measurement more versatile by reworking GCC + optional leaf into a chain of muxes.

cc @MarijnS95 @konradybcio-work

### CODE REVIEW

**Comment by lumag** [comment]
Note, this is an early draft. I'd like to discuss the concept before finishing the conversion. This was not validated on the hw, so it can be buggy even for the reworked platforms.

File: debugcc.h [hunk 1]
```diff
@@ -43,6 +43,9 @@ struct debug_mux {
 	const char *block_name;
 	size_t size;
 
+	struct debug_mux *parent;
+	unsigned long mux;
```
Comment (MarijnS95) [inline debugcc.h:47]: Nit: I feel like we should rename this to a perhaps-more-clear `parent_mux_value`? And perhaps have some doc-comments on the members?

**Review by MarijnS95 — APPROVED** [review]
Sounds like a good idea!

**Comment by lumag** [comment]
@MarijnS95 @konradybcio please review, I think I'm done with the implementation.

File: msm8936.c [hunk 2]
```diff
@@ -49,137 +49,139 @@
...
-	 { "gcc_mdss_byte1_clk", &gcc, 443 },
-	 { "gcc_mdss_esc0_clk", &gcc, 509 },
-	 { "gcc_mdss_esc1_clk", &gcc, 444 },
-	 { "gcc_bimc_clk", &gcc, 340 },
-	 { "gcc_bimc_gpu_clk", &gcc, 343 },
-	 { "gcc_bimc_ddr_ch0_clk", &gcc, 346 },
-	 { "gcc_cpp_tbu_clk", &gcc, 233 },
-	 { "gcc_mdp_rt_tbu_clk", &gcc, 238 },
-	 { "wcnss_m_clk", &gcc, 408 },
+	 { "gcc.mux_gp1_clk", &gcc.mux, 16 },
```
Comment (MarijnS95) [inline msm8936.c:77]: Did you intend to `s/gcc_/gcc./` in the string? I don't see this in other platforms.

Comment (lumag) [reply msm8936.c:77]: Stupid me and stupid regexps. I'll fix that.

File: msm8996.c [hunk 3]
```diff
@@ -86,6 +94,10 @@ static struct debug_mux apc0_mux = {
 	.phys = 0x06400000,
 	.size = 0x1000,
 
+	.measure = measure_leaf,
+	.parent = &gcc.mux,
+	.parent_mux_val = 0x00bb,
```
Comment (MarijnS95) [inline msm8996.c:99]: Question for the original code: why not just `0xbb`?

File: qcs404.c [hunk 4]
```diff
@@ -49,147 +50,190 @@
...
+static struct turing_mux turing = {
+	.mux = {
+		.phys = 0x800000,
+		.size = 0x30000,
 
-	.enable_reg = 0x22008,
-	.enable_mask = BIT(0),
+		.measure = measure_turing,
+		.parent = &gcc.mux,
+		.parent_mux_val = 50,
```
Comment (MarijnS95) [inline qcs404.c:94]: Issue with the original code: unfortunate that this isn't hex, while everything else is.

File: sc8280xp.c [hunk 5]
```diff
@@ -99,334 +111,334 @@ static struct debug_mux mc_cc = {
...
-	{ "disp1_cc_mdss_pclk1_clk",				&gcc, 0x82, &disp1_cc, 0x10 },
-	{ "disp1_cc_mdss_rot1_clk",				&gcc, 0x82, &disp1_cc, 0x14 },
-	{ "disp1_cc_mdss_rot_clk",				&gcc, 0x82, &disp1_cc, 0x13 },
-	{ "disp1_cc_mdss_rscc_ahb_clk",				&gcc, 0x82, &disp1_cc, 0x3c },
-	{ "disp1_cc_mdss_rscc_vsync_clk",			&gcc, 0x82, &disp1_cc, 0x3b },
-	{ "disp1_cc_mdss_vsync1_clk",				&gcc, 0x82, &disp1_cc, 0x18 },
-	{ "disp1_cc_mdss_vsync_clk",				&gcc, 0x82, &disp1_cc, 0x17 },
-	{ "disp1_cc_sleep_clk",					&gcc, 0x82, &disp1_cc, 0x46 },
-	{ "disp1_cc_xo_clk",					&gcc, 0x82, &disp1_cc, 0x45 },
-	{ "measure_only_mccc_clk", 				&gcc, 0xfeedbeef, &mc_cc, 0x50 },
```
Comment (MarijnS95) [inline sc8280xp.c:429]: Neat! Now `mc_cc` no longer has a fake/useless gcc parent and mux value!

File: sdm845.c [hunk 6]
```diff
@@ -146,252 +170,252 @@ static struct debug_mux video_cc = {
...
+	{ "gpu_cc_cx_snoc_dvm_clk", &gpu_cc, 22 },
+	{ "gpu_cc_sleep_clk", &gpu_cc, 23 },
+	{ "gpu_cc_cx_qdss_trig_clk", &gpu_cc, 24 },
+	{ "gpu_cc_cx_gmu_clk", &gpu_cc, 25 },
+	{ "gpu_cc_cx_gfx3d_clk", &gpu_cc, 26 },
+	{ "gpu_cc_cx_gfx3d_slv_clk", &gpu_cc, 27 },
+	{ "gpu_cc_rbcpr_clk", &gpu_cc, 28 },
+	{ "gpu_cc_rbcpr_ahb_clk", &gpu_cc, 29 },
+	{ "gpu_cc_acd_cxo_clk", &gpu_cc, 31 },
+	{ "gcc_gpu_memnoc_gfx_clk", &gcc.mux, 325 },
```
Comment (MarijnS95) [inline sdm845.c:411]: Very nicely sorted file :upside_down_face:

File: msm8996.c (hunk 3 above)
Comment (lumag) [reply msm8996.c:99]: Just because the table around it was using 4-difit values for the mux value. Fixed this chunk now.

File: sc8280xp.c (hunk 5 above)
Comment (lumag) [reply sc8280xp.c:429]: Yes, that was one of the main reasons!

File: sm6115.c [hunk 7]
```diff
@@ -81,131 +97,131 @@ static struct debug_mux gpu_cc = {
...
-	{ "gpu_cc_cx_snoc_dvm_clk", &gcc, 0xe7, &gpu_cc, 0x15 },
-	{ "gpu_cc_cxo_aon_clk", &gcc, 0xe7, &gpu_cc, 0xa },
-	{ "gpu_cc_cxo_clk", &gcc, 0xe7, &gpu_cc, 0x19 },
-	{ "gpu_cc_gx_cxo_clk", &gcc, 0xe7, &gpu_cc, 0xe },
-	{ "gpu_cc_gx_gfx3d_clk", &gcc, 0xe7, &gpu_cc, 0xb },
-	{ "gpu_cc_sleep_clk", &gcc, 0xe7, &gpu_cc, 0x16 },
+	{ "perfcl_clk", &cpu_cc, 0x1 },
+	{ "pwrcl_clk", &cpu_cc, 0x0 },
+	//{ "cpu_cc_debug_mux", &gcc.mux, 0xaf },
+	//{ "disp_cc_debug_mux", &gcc.mux, 0x42 },
```
Comment (MarijnS95) [inline sm6115.c:103]: IMHO these comments are even more useless now, can we just remove them?

File: qcs404.c (hunk 4 above)
Comment (lumag) [reply qcs404.c:94]: qcs404, sdm845 and msm8936 use decimal values instead of hex. I didn't touch that, since it is out-of-scope for this PR.

**Review by MarijnS95 — APPROVED** [review]
I haven't cross-checked every platform change, but the architectural rework and function implementations look fine to me!

Will try to test on a few platforms when I get the chance.

File: qcs404.c (hunk 4 above)
Comment (lumag) [reply qcs404.c:94]: Opened this as a separate issue (#27 )

File: sm6115.c (hunk 7 above)
Comment (lumag) [reply sm6115.c:103]: Yep, why not.

Comment (lumag) [reply sm6115.c]: Dropped most of useless comments, retained only those debug muxes that we do not implement for now.

**Comment by MarijnS95** [comment]
Quickly tested on `sm8150`, seems to work fine :+1:

**Comment by superna9999** [comment]
Tested on sm8550, works fine :+1:

**Comment by lumag** [comment]
My primary testing targets were msm8996 and sm8250.

**Pull request merged**